plotit/systematics.py

```diff
@@ -0,0 +1,200 @@
+    def _repr_args(self):
+        return (self.name, self.value)
+
+    def up(self, hist):
+        return hist.contents*self.value
+    def down(self, hist):
+        return hist.contents*(2-self.value)
+
+class LogNormalSystVar(ParameterizedSystVar):
+    """ """
+    def __init__(self, name, prior, postfit=0., postfit_error=None, postfit_error_up=None, postfit_error_down=None, pretty_name=None, on=SystVar.default_filter):
+        if postfit_error_up is None:
+            if postfit_error is not None:
+                postfit_error_up = postfit_error
+            else:
+                postfit_error_up = 1.
+        if postfit_error_down is None:
+            if postfit_error is not None:
+                postfit_error_down = postfit_error
+            else:
+                postfit_error_down = 1.
+        # eval
+        import math
+        self.value = math.exp(postfit*math.log(prior))
+        self.value_up = math.exp((postfit+postfit_error_up)*math.log(prior))
+        self.value_down = math.exp((postfit-postfit_error_down)*math.log(prior))
+        super(LogNormalSystVar, self).__init__(name, pretty_name=pretty_name, on=on)
+    ## TODO __repr__
+
+    def up(self, hist):
+        return hist.contents*self.value_up
+    def down(self, hist):
+        return hist.contents*self.value_down
+
+class ShapeSystVar(SystVar):
+    """ for shapes """
+    def __init__(self, name, pretty_name=None, on=SystVar.default_filter):
+        self.name = name
+        super(ShapeSystVar, self).__init__(name, pretty_name=pretty_name, on=on)
+
+    class ForHist(SystVar.ForHist):
+        __slots__ = ("_histUp", "_histDown")
+        def __init__(self, hist, systVar):
+            super(ShapeSystVar.ForHist, self).__init__(hist, systVar)
+            self._histUp = None
+            self._histDown = None
+        @lazyload
+        def histUp(self):
+            return self._findVarHist("up")
+        @lazyload
+        def histDown(self):
+            return self._findVarHist("down")
+        def _findVarHist(self, vari):
+            variHistName = "{0}__{1}{2}".format(self.hist.name, self.systVar.name, vari)
+            if self.hist.tFile.Get(variHistName):
+                return self.hist.clone(name=variHistName)
+            else: ## try to find the file
+                import os.path
+                fullpath = self.hist.tFile.GetPath().split(":")[0]
+                variPath = os.path.join(os.path.dirname(fullpath), "{0}__{1}{2}.root".format(os.path.splitext(os.path.basename(fullpath))[0], self.systVar.name, vari))
+                if os.path.exists(variPath):
+                    from cppyy import gbl
+                    vf = gbl.TFile.Open(variPath)
+                    if ( not vf ) or vf.IsZombie() or ( not vf.IsOpen() ):
+                        raise IOError("Could not open file '{}' correctly".format(variPath))
+                    if vf.Get(self.hist.name):
+                        return self.hist.clone(tfile=vf)
```

Comment:
This should be `tFile`

(bamboo crashed while trying to produce a cutflow report, I'm puzzled why it hadn't happened before...)

Reply:
hmm, probably because of a systematic that is taken from another file?

Reply:
I like the third option, it would also allow to use it elsewhere without having to build the "legacy" plotit. As for the name, I don't have a strong opinion, as long as there is no confusion with https://github.com/scikit-hep/mplhep ;) (it seems there is some feature overlap?)

I had in mind to try using it for making plots that differ from the "basic" data vs. MC stack paradigm of plotIt, e.g. for unfolded distributions with multiple ratios, but haven't looked at all into it. Using mplhep or just plain dumb ROOT scripts are also still options for that. As for the code, I doubt I'll have the time for any meaningful review, sorry...

Reply:
There is some overlap with mplhep indeed - but the parts that are done nicely in mplhep (styling etc.) are rather under-developed here, so we could use mplhep to implement the drawing with matplotlib (the overlap is mostly in mplbplot.draw_). The old way (mplbplot without this PR) to convert ROOT objects to arrays was with decorators (a `bins` generator), but the plotit reimplementation mostly uses direct conversion to numpy (which is a lot easier than before, and should be more efficient) internally, so this part should be revisited anyway.

Taking as an example the currrent matplotlib drawing code (since it's simpler - the ROOT version is a rather literal translation from the plotIt implementation), the actual drawing is done in [these lines](https://github.com/pieterdavid/mplbplot/blob/791feeda5e32475221a61ba90c01b8dad0f5bc51/plotit/draw_mpl.py#L81-L106). Given some instructions on how to build the different stacks, it should be easy to draw multiple (with ratios), e.g. by making an interface for what's now the StackRatioPlot class, with a method to customise constructing stacks from the list of samples, and for drawign them (the other styling+helpers can be in a base class). I have a few scripts that just

Reply:
I pushed a first version to https://gitlab.cern.ch/cp3-cms/pyplotit - basically a snapshot of this branch, but with a consistent code style (black) and pyupgrade to move to python 3.6+ (f-strings etc.) - testing still to be done. I included from mplbplot what is currently used, but as mentioned above it may be better to use mplhep instead. Feedback welcome, I'll try to do a bit of testing, write some basic documentation, and fill the issue tracker.

Reply:
Great, I'll take a look! :+1:

Reply:
I just pushed a few commits that make https://gitlab.cern.ch/cp3-cms/pyplotit work as a replacement for yields tables from bamboo, therefore this can be closed and other issues reported there (it also has [documentation](https://pyplotit.readthedocs.io/en/latest/index.html), which could become a useful reference for the plotIt file scheme).